### DISCUSSION TITLE
IoInput with char tokens

### DISCUSSION THREAD

**Discussion by judemille** [zesterer/chumsky#590] [Ideas]
I'm not entirely sure how this would be implemented. I would *like,* though, to be able to run chumsky over a file that should be UTF-8, and have it transparently handle incoming data as UTF-8, producing a non-recoverable error if it is invalid.

Given direction, I could help implement this.

**Comment by zesterer** [comment]
Hi.

I think it would be desirable to have something like this, yes. You can currently do it with `Stream::from_iter`, but one thing to consider is *backtracking*. Parsers in chumsky require backtracking support so `Stream` currently stores every token it's encountered so far so it can revisit them if necessary.

Obviously, this is undesirable: files can be far larger than the available memory so there needs to be a way to drop unnecessary input. One option might be to give parsers a way to say "I'll definitely never need anything before offset `x` from now on". For example, `Repeated` (which is pretty common as a top-level parser) along with a few others can all have this sort of thing encoded in them. This would allow us to pass this information back to the input, allowing it to drop tokens that it's finished performing work on.

Implementation-wise, I imagine this as a function on `Input` like

```rs
trait Input {
    ...
    /// If desirable, discard tokens with an offset less than `offset`. This is purely an optimisation: a no-op does not change parser semantics.
    /// The user of the input will promise to not query tokens before this offset after this call
    fn hint_discard_before(&self, offset: Self::Offset) {}
    ...
}
```

As an aside, `Stream` is long due a rework: it tries to do too many things at once and as a result is a little awkward to use, so please don't treat it as a golden goose. At minimum, I'd like to split it into maybe two or three other input types.

**Reply by judemille** [reply]
Yeah, the sorts of formats I'm reading are currently line-based, and even for the ones that hold a few lines' worth of context, there is still a predictable point at which I can say "okay, I don't need the previous input anymore".
